### PR TITLE
UAF-3159 / UAF-31 Check Reconciliation module

### PR DESCRIPTION
**Pull request by adamkost**
Code Feature for the Check Reconciliation module UAF-3159

Functional specs available at https://jira.arizona.edu/browse/UAF-3159

### CODE REVIEW

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+                LOG.info("Processing file " + file.getAbsolutePath());
+                deleteDoneFile(file);
+                processFile(file);
+            } catch (BankNotFoundException e) {
+                // swallow the exceptions, and try to parse the other files
+                LOG.error("Bank account number could not be matched in: " + file.getAbsolutePath() + ". This file's processing has been rolledback.", e);
+                writeDoneFile(file);
+            } catch (Exception e) {
+                LOG.error("Exception occured during loading of " + file.getAbsolutePath() + ". This file's processing has been rolledback.", e);
+                success = false;
```
Comment (sskinner) [inline kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java]: No-op, can be deleted.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+                throw new Exception(e);
+            } finally {
```
Comment (sskinner) [inline kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java]: No-op, can deleted.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+        try {
+            FileUtils.deleteQuietly(new File(doneFile));
+        } catch (Exception e) {
+            LOG.info("Error in CheckReconciliationBatchService in method deleteDoneFile");
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Transactional
+    private void processFile(File file) throws Exception {
```
Comment (sskinner) [inline kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java:230]: Please ensure a finally block closes the BufferedReader.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            accountTotal += reportLine.getAmount();
+            accountCount++;
+            overallTotal += reportLine.getAmount();
+            overallCount++;
+        }
+
+        out.println("Sub-total: " + currentMonth + KFSConstants.COMMA + KFSConstants.COMMA + "Outstanding: " + monthCount + KFSConstants.COMMA + KFSConstants.COMMA + KFSConstants.COMMA + CrConstants.REPORT_DECIMAL_FORMAT.format(monthTotal));
+        out.println("Bank: " + currentAccount + KFSConstants.COMMA + KFSConstants.COMMA + "Total Outstanding: " + accountCount + KFSConstants.COMMA + KFSConstants.COMMA + KFSConstants.COMMA + CrConstants.REPORT_DECIMAL_FORMAT.format(accountTotal));
+
+        out.flush();
```
Comment (sskinner) [inline kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java]: Let's make sure to use 'finally' to close resources.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+        ResourceBundle resourceBundle = ResourceBundle.getBundle(CrConstants.REPORT_MESSAGES_CLASSPATH, Locale.getDefault());
+        String date = CrConstants.MM_DD_YYYY.format(getReportAsOfDate(reportEndDate));
+        Map<String, Object> reportData = new HashMap<String, Object>();
+        reportData.put(JRParameter.REPORT_RESOURCE_BUNDLE, resourceBundle);
+        reportData.put(ReportGeneration.PARAMETER_NAME_SUBREPORT_DIR, KFSConstants.EMPTY_STRING);
+        reportData.put(CrPropertyConstants.CheckReconciliationReport.END_DATE, date);
+
+        reportGenerationService.generateReportToOutputStream(reportData, reportSet, CrConstants.REPORT_TEMPLATE_CLASSPATH, baos);
+        WebUtils.saveMimeOutputStreamAsFile(response, KFSConstants.ReportGeneration.PDF_MIME_TYPE, baos, CrConstants.CR_REPORT_FILE_NAME + KFSConstants.ReportGeneration.PDF_FILE_EXTENSION);
```
Comment (sskinner) [inline kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java]: A try/finally right here for baos?

**Review by sskinner — CHANGES_REQUESTED** [review]
Great job! Awesome first review, little nits to address, and we'll get the merge going.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java (hunk 1 above)
Comment (adamkost) [reply kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java]: Fixed.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java (hunk 2 above)
Comment (adamkost) [reply kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java]: Fixed.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java (hunk 3 above)
Comment (adamkost) [reply kfs-cr/src/main/java/edu/arizona/kfs/module/cr/batch/service/impl/CheckReconciliationImportBatchServiceImpl.java:230]: Finally and additional try blocks added at potential exception points, closing the BufferedReader before throwing exceptions.

File: kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java (hunk 4 above)
Comment (adamkost) [reply kfs-cr/src/main/java/edu/arizona/kfs/module/cr/service/impl/CheckReconciliationReportServiceImpl.java]: Fixed.

**Comment by adamkost** [comment]
changes ready for review.

**Review by sskinner — APPROVED** [review]
Looks awesome, thanks for the quick turn-around!

**Pull request merged**